File: Greedy/EqualSumArraysWithMinimumNumberOfOperations.cpp
//Problem Link:- https://leetcode.com/problems/equal-sum-arrays-with-minimum-number-of-operations/

#include<bits/stdc++.h>
using namespace std;

class Solution {
public:
    int minOperations(vector<int>& nums1, vector<int>& nums2) {
        int first = accumulate(nums1.begin(), nums1.end(),0);
        int sec = accumulate(nums2.begin(), nums2.end(),0);
        vector<int> arr1;
        vector<int> arr2;
        if (first == sec)
            return 0;
        if (sec > first){
            arr1 = nums2;
            arr2 = nums1;
        }
        else{
            arr1 = nums1;
            arr2 = nums2;
        }
        cout <<arr1.size() << endl;
        priority_queue<int> pq;
        for(int i=0;i<arr1.size();i++){
            pq.push(arr1[i]-1);
        }
        for(int i=0;i<arr2.size();i++){
            pq.push(6-arr2[i]);
        }
        int cnt = 0;
        int val;
        int diff = abs(first - sec);
        while (pq.size() > 0 && diff > 0){
            val = pq.top();
            pq.pop();
            diff -= val;
            cnt++;
        } 
        if (diff > 0)
            return -1;
        return cnt;
    }
};

File: Greedy/EqualSumArraysWithMinimumNumberOfOperations.py
#Problem Link:- https://leetcode.com/problems/equal-sum-arrays-with-minimum-number-of-operations/

"""You are given two arrays of integers nums1 and nums2, possibly of different lengths. The values 
in the arrays are between 1 and 6, inclusive.In one operation, you can change any integer's value 
in any of the arrays to any value between 1 and 6, inclusive.Return the minimum number of operations 
required to make the sum of values in nums1 equal to the sum of values in nums2. Return -1​​​​​ if it is 
not possible to make the sum of the two arrays equal."""

class Solution(object):
    def minOperations(self, nums1, nums2):
        import heapq
        first = sum(nums1)
        sec = sum(nums2)
        if sec == first:
            return 0
        if sec > first:
            nums1,nums2 = nums2,nums1
            first,sec = sec,first
        heap = []
        for i in range(len(nums1)):
            heappush(heap,-(nums1[i] - 1))
        for i in range(len(nums2)):
            heappush(heap,-(6 - nums2[i]))
        diff = abs(first-sec)
        cnt = 0
        while heap and diff > 0:
            val = -heappop(heap)
            diff -= val
            cnt += 1
        if diff > 0:
            return -1
        return cnt